src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -1075,6 +1075,92 @@ bool SystemDictionary::check_shared_class_super_types(InstanceKlass* ik, Handle
   return true;
 }
 
+// Pre-load class referred to in non-static null-free instance field. These fields trigger MANDATORY loading
+bool SystemDictionary::preload_from_null_free_field(InstanceKlass* ik, Handle class_loader, Symbol* sig, int field_index, TRAPS) {
+  TempNewSymbol name = Signature::strip_envelope(sig);
+  log_info(class, preload)("Preloading class %s during loading of shared class %s. \
+                           Cause: a null-free non-static field is declared with this type",
+                           name->as_C_string(), ik->name()->as_C_string());
+  Klass* real_k = SystemDictionary::resolve_with_circularity_detection_or_fail(ik->name(), name,
+                                                                               class_loader, false, CHECK_false);
+  if (HAS_PENDING_EXCEPTION) {
+    log_warning(class, preload)("Preloading of class %s during loading of class %s \
+                                (cause: null-free non-static field) failed: %s",
+                                name->as_C_string(), ik->name()->as_C_string(),
+                                PENDING_EXCEPTION->klass()->name()->as_C_string());
+    return false; // Exception is still pending
+  }
+
+  Klass* k = ik->get_inline_type_field_klass_or_null(field_index);
+  if (real_k != k) {
+    // oops, the app has substituted a different version of k!
+    log_warning(class, preload)("Preloading of class %s during loading of shared class %s \
+                                (cause: null-free non-static field) failed : \
+                                app substituted a different version of %s",
+                                name->as_C_string(), ik->name()->as_C_string(),
+                                k->name()->as_C_string());
+    return false;
+  }
+  log_info(class, preload)("Preloading of class %s during loading of shared class %s \
+                           (cause: null-free non-static field) succeeded",
+                           name->as_C_string(), ik->name()->as_C_string());
+
+  assert(real_k != nullptr, "Sanity check");
+  if (real_k->access_flags().is_identity_class()) {
```

Comment:
It looks like you could refactor the check for identity and abstract in to InstanceKlass::check_inline_field(TRAPS) and share the code with ClassFileParser. (?)

Reply:
This check is being done in Klass so I would need to add it there. I think a better name might be `Klass::check_null_restricted_field(TRAPS)` since these are mandatory conditions for a field to be null restricted. The checks may actually be unnecessary since the `@NullRestricted` annotation can only be used on a concrete value field.

Reply:
It should be InstanceKlass not Klass though since Klass has no fields.  resolve_with_circularity_check does return an InstanceKlass so the return type should be changed to InstanceKlass.  The check_null_restricted_field name seems good though.